### PR TITLE
calculus revisions

### PR DESCRIPTION
**Pull request by zackchase**
Major polish pass through the calculus notebook.

By submitting this pull request, I confirm that you can use, modify,
copy, and redistribute this contribution, under the terms of your
choice.

### CODE REVIEW

**Comment by mli** [comment]
Job d2l-en/PR-1858/3 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1858/

**Comment by mli** [comment]
Job d2l-en/PR-1858/4 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1858/

**Comment by mli** [comment]
Job d2l-en/PR-1858/6 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1858/

File: chapter_preliminaries/calculus.md [hunk 1]
```diff
@@ -1,41 +1,97 @@
...
 by an infinitesimally small amount.
-
-Suppose that we have a function $f: \mathbb{R} \rightarrow \mathbb{R}$,
-whose input and output are both scalars.
-[**The *derivative* of $f$ is defined as**]
+Formally, for functions $f: \mathbb{R} \rightarrow \mathbb{R}$,
+that map from scalars to scalars,
+[**the *derivative* of $f$ at a point $x$ is defined as**]
 
 (**$$f'(x) = \lim_{h \rightarrow 0} \frac{f(x+h) - f(x)}{h},$$**)
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:56]: , -> .

File: chapter_preliminaries/calculus.md [hunk 2]
```diff
@@ -1,41 +1,97 @@
...
-If $f$ is differentiable everywhere on a set, say $[a,b]$, 
-then $f$ is referred to as differentiable on this set.
-We can interpret the derivative $f'(x)$ in :eqref:`eq_derivative`
-as the *instantaneous* rate of change of $f(x)$
-with respect to $x$.
-For more intuition let us experiment with an example.
+This term on the right hand side is called a *limit* 
+and it tells us what happens 
+to the value of an expression
+(doesit blow up to infinity
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:62]: does it ?

File: chapter_preliminaries/calculus.md [hunk 3]
```diff
@@ -1,41 +1,97 @@
...
-then $f$ is referred to as differentiable on this set.
-We can interpret the derivative $f'(x)$ in :eqref:`eq_derivative`
-as the *instantaneous* rate of change of $f(x)$
-with respect to $x$.
-For more intuition let us experiment with an example.
+This term on the right hand side is called a *limit* 
+and it tells us what happens 
+to the value of an expression
+(doesit blow up to infinity 
+or converge to a some value)
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:63]: do we need a ?

**Comment by mli** [comment]
Job d2l-en/PR-1858/8 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1858/

File: chapter_preliminaries/calculus.md [hunk 4]
```diff
@@ -1,41 +1,97 @@
...
+as a specified variable 
+approaches a particular value.
+This limit tells us what 
+the ratio between a perturbation $h$
+and the change in the function value 
+$f(x + h) - f(x)$ converges to 
+as we shrink its size to zero.
+
+When $f'(x)$ exists, $f$ is said 
+to be *differentiable* at $a$.
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:73]: at $x$

File: chapter_preliminaries/calculus.md [hunk 5]
```diff
@@ -1,41 +1,97 @@
...
+And when $f'(x)$ exists for all $x$
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:74]: Can we start a sentence with "And"?

File: chapter_preliminaries/calculus.md [hunk 6]
```diff
@@ -1,41 +1,97 @@
...
+
+
+We can interpret the derivative 
+$f'(x)$ in :eqref:`eq_derivative`
+as the *instantaneous* rate of change 
+of $f(x)$ with respect to $x$.
+
+
+
+Let's develop some intuition with an example.
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:94]: combine this paragraph with the previous one?

File: chapter_preliminaries/calculus.md [hunk 7]
```diff
@@ -71,11 +127,10 @@ def f(x):
...
-[**By setting $x=1$ and letting $h$ approach $0$,
-the numerical result of $\frac{f(x+h) - f(x)}{h}$**]
-in :eqref:`eq_derivative`
-(**approaches $2$.**)
-Though this experiment is not a mathematical proof,
+[**Setting $x=1$, $\frac{f(x+h) - f(x)}{h}$**] (**approaches $2$
+as as $h$ approach $0$.**)
+While this experiment lacks 
+the rigor of a mathematical proof,
 we will see later see that indeed $f'(1) = 2$.
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:134]: we will later see

File: chapter_preliminaries/calculus.md [hunk 8]
```diff
@@ -84,28 +139,46 @@ for h in 10.0**np.arange(-1, -6, -1):
...
-We can use the following rules to differentiate common functions:
+where the symbols $\frac{d}{dx}$ and $D$ are *differentiation operators*.
+Below, we present the derivatives of some common functions:
 
-$$\begin{aligned} \frac{d}{dx} C & = 0 && \text{$C$ is a constant} \\ \frac{d}{dx} x^n & = n x^{n-1} && \text{for } n \neq 0 \\ \frac{d}{dx} e^x & = e^x \\ \frac{d}{dx} \ln x & = x^{-1} \end{aligned}$$
-
-To differentiate functions obtained by composition of the above (and similar functions) the following rules are handy. We assume below that $f$ and $g$ are both differentiable and that $c$ is constant. We have
+$$\begin{aligned} \frac{d}{dx} C & = 0 && \text{for any constant $C$} \\ \frac{d}{dx} x^n & = n x^{n-1} && \text{for } n \neq 0 \\ \frac{d}{dx} e^x & = e^x \\ \frac{d}{dx} \ln x & = x^{-1} \end{aligned}$$
 
+Functions composes from differentiable functions
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:152]: composed

File: chapter_preliminaries/calculus.md [hunk 9]
```diff
@@ -84,28 +139,46 @@ for h in 10.0**np.arange(-1, -6, -1):
...
-
-To differentiate functions obtained by composition of the above (and similar functions) the following rules are handy. We assume below that $f$ and $g$ are both differentiable and that $c$ is constant. We have
+$$\begin{aligned} \frac{d}{dx} C & = 0 && \text{for any constant $C$} \\ \frac{d}{dx} x^n & = n x^{n-1} && \text{for } n \neq 0 \\ \frac{d}{dx} e^x & = e^x \\ \frac{d}{dx} \ln x & = x^{-1} \end{aligned}$$
 
+Functions composes from differentiable functions 
+are often themselves differentiable.
+The following rules come in handy 
+for working with compositions 
+of any differentiable functions 
+$f$ and $g$, and constant $c$.
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:157]: $C$

File: chapter_preliminaries/calculus.md [hunk 10]
```diff
@@ -84,28 +139,46 @@ for h in 10.0**np.arange(-1, -6, -1):
...
 $$\begin{aligned} \frac{d}{dx} [C f(x)] & = C \frac{d}{dx} f(x) && \text{Constant multiple rule} \\ \frac{d}{dx} [f(x) + g(x)] & = \frac{d}{dx} f(x) + \frac{d}{dx} g(x) && \text{Sum rule} \\ \frac{d}{dx} [f(x) g(x)] & = f(x) \frac{d}{dx} g(x) + g(x) \frac{d}{dx} f(x) && \text{Product rule} \\ \frac{d}{dx} \frac{f(x)}{g(x)} & = \frac{g(x) \frac{d}{dx} f(x) - f(x) \frac{d}{dx} g(x)}{g^2(x)} && \text{Quotient rule} \end{aligned}$$
 
-Using this we can apply the rules to find the derivative of $3 x^2 - 4x$ via
+Using this, we can apply the rules 
+to find the derivative of $3 x^2 - 4x$ via
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:161]: apply xxx to find 
or 
apply xxx to finding?

File: chapter_preliminaries/calculus.md [hunk 11]
```diff
@@ -84,28 +139,46 @@ for h in 10.0**np.arange(-1, -6, -1):
...
 
-[**To visualize slopes defined by derivatives we will use `matplotlib`,**] a popular plotting library in Python.
-We need to define a few functions. As its name indicates, `use_svg_display` tells `matplotlib` to output graphics in SVG format for crisper images. The comment `#@save` is a special modifier that allows us to save the subsequent function, class, or statements in the `d2l` package such that it can be invoked later without being redefined, e.g., via `d2l.use_svg_display()`. 
+[**We can visualize the slopes of functions using the `matplotlib` library**].
+We need to define a few functions. 
+As its name indicates, `use_svg_display` 
+tells `matplotlib` to output graphics 
+in SVG format for crisper images. 
+The comment `#@save` is a special modifier 
+that allows us to save the subsequent function,
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:177]: the -> any

File: chapter_preliminaries/calculus.md [hunk 12]
```diff
@@ -183,72 +261,150 @@ plot(x, [f(x), 2 * x - 3], 'x', 'f(x)', legend=['f(x)', 'Tangent line (x=1)'])
...
-has variables $x_1, x_2, \ldots, x_n$, in short 
-$\mathbf{u} = g(\mathbf{x})$. Then the chain rule reads as follows:
+$u_1, u_2, \ldots, u_m$, 
+where each $u_i = g_i(\mathbf{x})$ 
+has variables $x_1, x_2, \ldots, x_n$,
+i.e.,  $\mathbf{u} = g(\mathbf{x})$.
+Then the chain rule states that
 
 $$\frac{dy}{dx_i} = \frac{dy}{du_1} \frac{du_1}{dx_i} + \frac{dy}{du_2} \frac{du_2}{dx_i} + \cdots + \frac{dy}{du_m} \frac{du_m}{dx_i} \text{ and thus }
 \nabla_{\mathbf{x}} y = \nabla_{\mathbf{u}} y \cdot \nabla_{x} \mathbf{u}$$
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:347]: \nabla_{x} -> \nabla_{\mathbf{x}}

File: chapter_preliminaries/calculus.md [hunk 13]
```diff
@@ -183,72 +261,150 @@ plot(x, [f(x), 2 * x - 3], 'x', 'f(x)', legend=['f(x)', 'Tangent line (x=1)'])
...
 
-Note that $\nabla_{x} \mathbf{u}$ is a *matrix* since it contains the derivative of a vector with regard to a vector. As such, evaluating the gradient requires us to compute a vector-matrix product. This is one of the key reasons why linear algebra is such an integral building block in building deep learning systems.
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:232]: \nabla_{x} -> \nabla_{\mathbf{x}}

File: chapter_preliminaries/calculus.md [hunk 14]
```diff
@@ -183,72 +261,150 @@ plot(x, [f(x), 2 * x - 3], 'x', 'f(x)', legend=['f(x)', 'Tangent line (x=1)'])
...
+
+
+
+## Discussion
+
+While we have just scratched the surface of a deep topic,
+a number of concepts already come into focus: 
+first, the composition rules for differentiation 
+allow us to compute the derivatives of complex functions
+*automatically* by rote application.
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:366]: rote learning?

File: chapter_preliminaries/calculus.md [hunk 15]
```diff
@@ -183,72 +261,150 @@ plot(x, [f(x), 2 * x - 3], 'x', 'f(x)', legend=['f(x)', 'Tangent line (x=1)'])
...
+This task requires no creativity and thus 
+we can focus our cognitive powers elsewhere.
+Second, computing the derivatives of vector-valued functions 
+requires us to multiply matrices as we trace 
+the dependency graph of variables from output to input. 
+In particular, this graph is traversed in a *forward* direction 
+when we evaluate a function 
+and in a *backwards* direction 
+when we compute gradients. 
+Later chapters will formalize introduce backpropagation,
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:376]: formally

File: chapter_preliminaries/calculus.md [hunk 16]
```diff
@@ -183,72 +261,150 @@ plot(x, [f(x), 2 * x - 3], 'x', 'f(x)', legend=['f(x)', 'Tangent line (x=1)'])
...
+in order to lower the loss,
+and each step of the optimization algorithms used 
+throughout this book will require calculating the gradient.
 
 ## Exercises
 
-1. So far we took the rules for derivatives for granted. Using the definition and limits prove the properties for a) $f(x) = c$, b) $f(x) = x^n$, c) $f(x) = e^x$ and d) $f(x) = \ln x$.
+1. So far we took the rules for derivatives for granted. 
+   Using the definition and limits prove the properties 
+   for a) $f(x) = c$, b) $f(x) = x^n$, c) $f(x) = e^x$ and d) $f(x) = \ln x$.
```
Comment (astonzhang) [inline chapter_preliminaries/calculus.md:389]: (i) (ii) (iii) (iv)

Comment (astonzhang) [reply chapter_preliminaries/calculus.md:389]: \log

File: chapter_preliminaries/autograd.md [hunk 17]
```diff
@@ -1,43 +1,62 @@
...
-description of backpropagation of the form that is used in deep learning systems did not 
-happen until the PhD thesis of :cite:`speelpenning1980compiling`. More modern descriptions 
-can be found in the work of :cite:`griewank1989automatic` who put autograd on a comprehensive footing. 
-While by now backpropagation is the default choice in computing gradients, it is by far not the only option. For instance Julia employs forward propagation :cite:`revels2016forward`. We will discuss this in some more detail after we've introduced the basics.
+Recall from :numref:`sec_calculus` 
+that calculating derivatives is the crucial step
+in all of the optimization algorithms
+that we will use to train deep networks.
+While the calculations are straightforward,
+working them out by hand can tedious and error-prone,
```
Comment (astonzhang) [inline chapter_preliminaries/autograd.md:9]: can be

**Comment by mli** [comment]
Job d2l-en/PR-1858/9 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1858/

File: chapter_preliminaries/autograd.md [hunk 18]
```diff
@@ -342,7 +461,14 @@ def f(a):
     return c
 ```
 
-Let us compute the gradient. Since the input is a random variable, we cannot predict the flow a priori. It is only by executing `f(a)` that we know which operations will be carried out, and how frequently. 
+Below we call this function with a random variable as input
```
[the remark this answers is not in the document]
Comment (astonzhang) [reply chapter_preliminaries/autograd.md:464]: Below, we call this function, passing in a random value as input.

**Comment by mli** [comment]
Job d2l-en/PR-1858/10 is complete. 
Check the results at http://preview.d2l.ai/d2l-en/PR-1858/

**Pull request merged**